pkg/controller/packetsampling/controller.go

```diff
@@ -0,0 +1,363 @@
+
+	for _, psName := range ss {
+		// Re-post all running PacketSampling requests to the work queue to
+		// be processed and checked for timeout.
+		c.queue.Add(psName)
+	}
+}
+
+func (c *Controller) worker() {
+	for c.processPacketSamplingItem() {
+
+	}
+}
+
+func (c *Controller) startPacketSampling(ps *crdv1alpha1.PacketSampling) error {
+	tag, err := c.allocateTag(ps.Name)
+	if err != nil {
+		return err
+	}
+	if tag == 0 {
+		return nil
+	}
+
+	err = c.updatePacketSamplingStatus(ps, crdv1alpha1.PacketSamplingRunning, "", tag)
+	if err != nil {
+		c.deallocateTag(ps.Name, tag)
+	}
+
+	return err
+}
+
+func (c *Controller) updatePacketSamplingStatus(ps *crdv1alpha1.PacketSampling, phase crdv1alpha1.PacketSamplingPhase, reason string, dataPlaneTag uint8) error {
+	update := ps.DeepCopy()
+	update.Status.Phase = phase
+	if phase == crdv1alpha1.PacketSamplingRunning && update.Status.StartTime == nil {
+		t := metav1.Now()
+		update.Status.StartTime = &t
+	}
+	update.Status.DataplaneTag = int8(dataPlaneTag)
+	if reason != "" {
+		update.Status.Reason = reason
+	}
+
+	_, err := c.client.CrdV1alpha1().PacketSamplings().UpdateStatus(context.TODO(), update, metav1.UpdateOptions{})
+	return err
+}
+
+// Allocates a tag. If the PacketSampling request has been allocated with a tag
+// already, 0 is returned. If number of existing PacketSampling requests reaches
+// the upper limit, an error is returned.
+func (c *Controller) allocateTag(name string) (uint8, error) {
+	c.runningPacketSamplingsMutex.Lock()
+	defer c.runningPacketSamplingsMutex.Unlock()
+
+	for _, n := range c.runningPacketSamplings {
+		if n == name {
+			// The packetsampling request has been processed already.
+			return 0, nil
+		}
+	}
+	for i := minTagNum; i <= maxTagNum; i += tagStep {
+		if _, ok := c.runningPacketSamplings[i]; !ok {
+			c.runningPacketSamplings[i] = name
+			return i, nil
+		}
+	}
+	return 0, fmt.Errorf("number of on-going packetsampling operations already reached the upper limit: %d", maxTagNum)
+}
+
+func (c *Controller) processPacketSamplingItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.queue.Done(obj)
+
+	key, ok := obj.(string)
+	if !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	}
+
+	err := c.syncPacketSampling(key)
+
+	if err != nil {
+		klog.Errorf("error sync packetSampling %s, existing,error: %v", key, err)
+		c.queue.AddRateLimited(key)
+	} else {
+		c.queue.Forget(key)
+	}
+
+	return true
+}
+
+func (c *Controller) deallocateTagForPS(ps *crdv1alpha1.PacketSampling) {
+	if ps.Status.DataplaneTag != 0 {
+		c.deallocateTag(ps.Name, uint8(ps.Status.DataplaneTag))
+	}
+}
+
+func (c *Controller) deallocateTag(name string, tag uint8) {
+	c.runningPacketSamplingsMutex.Lock()
+	defer c.runningPacketSamplingsMutex.Unlock()
+
+	if existing, ok := c.runningPacketSamplings[tag]; ok {
+		if name == existing {
+			delete(c.runningPacketSamplings, tag)
+		}
```

Comment:
Usually we name it `exists`.
```suggestion
	if exists, ok := c.runningPacketSamplings[tag]; ok {
		if name == exists {
			delete(c.runningPacketSamplings, tag)
		}
```